File: server/api/v1/business/topology.go
package business

import (
	"github.com/flipped-aurora/gin-vue-admin/server/model/business"
	"github.com/flipped-aurora/gin-vue-admin/server/model/common/request"
	"github.com/flipped-aurora/gin-vue-admin/server/model/common/response"
	"github.com/gin-gonic/gin"
)

type TopologyApi struct{}

// Layer APIs
func (b *TopologyApi) CreateLayer(c *gin.Context) {
	var layer business.TopologyLayer
	err := c.ShouldBindJSON(&layer)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.CreateLayer(&layer)
	if err != nil {
		response.FailWithMessage("创建失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(layer, "创建成功", c)
}

func (b *TopologyApi) UpdateLayer(c *gin.Context) {
	var layer business.TopologyLayer
	err := c.ShouldBindJSON(&layer)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.UpdateLayer(&layer)
	if err != nil {
		response.FailWithMessage("更新失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(layer, "更新成功", c)
}

func (b *TopologyApi) DeleteLayer(c *gin.Context) {
	var reqId request.GetById
	err := c.ShouldBindJSON(&reqId)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.DeleteLayer(reqId.Uint())
	if err != nil {
		response.FailWithMessage("删除失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("删除成功", c)
}

func (b *TopologyApi) GetLayers(c *gin.Context) {
	layers, err := TopologyServiceApp.GetAllLayers()
	if err != nil {
		response.FailWithMessage("获取失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(layers, "获取成功", c)
}

// Category APIs
func (b *TopologyApi) CreateCategory(c *gin.Context) {
	var cat business.TopologyCategory
	err := c.ShouldBindJSON(&cat)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.CreateCategory(&cat)
	if err != nil {
		response.FailWithMessage("创建失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(cat, "创建成功", c)
}

func (b *TopologyApi) UpdateCategory(c *gin.Context) {
	var cat business.TopologyCategory
	err := c.ShouldBindJSON(&cat)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.UpdateCategory(&cat)
	if err != nil {
		response.FailWithMessage("更新失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(cat, "更新成功", c)
}

func (b *TopologyApi) DeleteCategory(c *gin.Context) {
	var reqId request.GetById
	err := c.ShouldBindJSON(&reqId)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.DeleteCategory(reqId.Uint())
	if err != nil {
		response.FailWithMessage("删除失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("删除成功", c)
}

func (b *TopologyApi) GetCategoriesByLayer(c *gin.Context) {
	layerID := c.Query("layer_id")
	categories, err := TopologyServiceApp.GetCategoriesByLayerID(parseUint(layerID))
	if err != nil {
		response.FailWithMessage("获取失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(categories, "获取成功", c)
}

// Vendor APIs
func (b *TopologyApi) CreateVendor(c *gin.Context) {
	var vendor business.TopologyVendor
	err := c.ShouldBindJSON(&vendor)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.CreateVendor(&vendor)
	if err != nil {
		response.FailWithMessage("创建失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(vendor, "创建成功", c)
}

func (b *TopologyApi) UpdateVendor(c *gin.Context) {
	var vendor business.TopologyVendor
	err := c.ShouldBindJSON(&vendor)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.UpdateVendor(&vendor)
	if err != nil {
		response.FailWithMessage("更新失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(vendor, "更新成功", c)
}

func (b *TopologyApi) DeleteVendor(c *gin.Context) {
	var reqId request.GetById
	err := c.ShouldBindJSON(&reqId)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = TopologyServiceApp.DeleteVendor(reqId.Uint())
	if err != nil {
		response.FailWithMessage("删除失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("删除成功", c)
}

func (b *TopologyApi) GetVendorsByCategory(c *gin.Context) {
	categoryID := c.Query("category_id")
	vendors, err := TopologyServiceApp.GetVendorsByCategoryID(parseUint(categoryID))
	if err != nil {
		response.FailWithMessage("获取失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(vendors, "获取成功", c)
}

File: web/src/api/business/topology.js
import service from '@/utils/request'

// ===================== Layer APIs =====================
// @Tags Topology
// @Summary 创建拓扑层级
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.TopologyLayer true "创建拓扑层级"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/layer [post]
export const createLayer = (data) => {
  return service({
    url: '/topology/layer',
    method: 'post',
    data
  })
}

// @Tags Topology
// @Summary 更新拓扑层级
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.TopologyLayer true "更新拓扑层级"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/layer [put]
export const updateLayer = (data) => {
  return service({
    url: '/topology/layer',
    method: 'put',
    data
  })
}

// @Tags Topology
// @Summary 删除拓扑层级
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body dbModel.TopologyLayer true "删除拓扑层级"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/layer [delete]
export const deleteLayer = (data) => {
  return service({
    url: '/topology/layer',
    method: 'delete',
    data
  })
}

// @Tags Topology
// @Summary 获取所有拓扑层级
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/layer/all [get]
export const getLayers = () => {
  return service({
    url: '/topology/layer/all',
    method: 'get'
  })
}

// ===================== Category APIs =====================
// @Tags Topology
// @Summary 创建拓扑分类
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.TopologyCategory true "创建拓扑分类"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/category [post]
export const createCategory = (data) => {
  return service({
    url: '/topology/category',
    method: 'post',
    data
  })
}

// @Tags Topology
// @Summary 更新拓扑分类
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.TopologyCategory true "更新拓扑分类"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/category [put]
export const updateCategory = (data) => {
  return service({
    url: '/topology/category',
    method: 'put',
    data
  })
}

// @Tags Topology
// @Summary 删除拓扑分类
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body dbModel.TopologyCategory true "删除拓扑分类"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/category [delete]
export const deleteCategory = (data) => {
  return service({
    url: '/topology/category',
    method: 'delete',
    data
  })
}

// @Tags Topology
// @Summary 按层级获取拓扑分类
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data query layerId true "层级ID"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/category/by-layer [get]
export const getCategoriesByLayer = (params) => {
  return service({
    url: '/topology/category/by-layer',
    method: 'get',
    params
  })
}

// ===================== Vendor APIs =====================
// @Tags Topology
// @Summary 创建拓扑厂商
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.TopologyVendor true "创建拓扑厂商"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/vendor [post]
export const createVendor = (data) => {
  return service({
    url: '/topology/vendor',
    method: 'post',
    data
  })
}

// @Tags Topology
// @Summary 更新拓扑厂商
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body business.TopologyVendor true "更新拓扑厂商"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/vendor [put]
export const updateVendor = (data) => {
  return service({
    url: '/topology/vendor',
    method: 'put',
    data
  })
}

// @Tags Topology
// @Summary 删除拓扑厂商
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body dbModel.TopologyVendor true "删除拓扑厂商"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/vendor [delete]
export const deleteVendor = (data) => {
  return service({
    url: '/topology/vendor',
    method: 'delete',
    data
  })
}

// @Tags Topology
// @Summary 按分类获取拓扑厂商
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data query categoryId true "分类ID"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/vendor/by-category [get]
export const getVendorsByCategory = (params) => {
  return service({
    url: '/topology/vendor/by-category',
    method: 'get',
    params
  })
}

// ===================== Public APIs (No Auth Required) =====================
// @Tags Topology
// @Summary 获取完整拓扑结构（供首页全景图使用）
// @accept application/json
// @Produce application/json
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /topology/full [get]
export const getFullTopology = () => {
  return service({
    url: '/topology/full',
    method: 'get'
  })
}
